Packet Calc Improvements

Bertie690 opened the pull request:
This contains all of my previous typo & bugfixing commits (as of Wednesday) as well as the new mineralpacket.estimateDamage function to simulate damage of an impending packet collision (and associated test func)
(this is literally just the other repo but with effectively only the packet related feature changes as I want to make sure those work before doing anything else)

Bertie690 committed:
Update message.go

Bertie690 committed:
Fixed typos in message.go

Still need to check other files with messages

Bertie690 committed:
Minor typo fixes

Bertie690 committed:
Minor typo fixes

adding missing periods, punctuation fixes, etc.

Bertie690 committed:
Merge branch 'sirgwain:develop' into develop

Bertie690 committed:
Update LRT Descriptions

Minor tweaks & clarifications

Bertie690 committed:
Update LRTsDescriptions.svelte

Typo fix

Bertie690 committed:
Update PRT Descriptions

Effectively resolving my recent issue by myself by updating the PRT description text to include additional info & details

Bertie690 committed:
Update LRTsDescriptions.svelte

Slight additional typo fixes

Bertie690 committed:
Update CE description

forgot to mention that it increases starting prop lvl by 1

Bertie690 committed:
Planet message typofixes

Minor updates (adding missing periods, etc.)

Bertie690 committed:
typo fixes

also made starbase ship build notif distinguish between singular/plural (hopefully shouldn't bork)

Bertie690 committed:
minor message tweaks

added #s for amount of ships involved to battle message descriptions (hopefully won't bork)

Bertie690 committed:
Update PlayerMessageDetail.svelte

minor tweaks

Bertie690 committed:
Update message.go

Bertie690 committed:
minor bugfix to traits.go

fixed a few bugs & starting pop errors

Bertie690 committed:
Improved capacitor/beam deflector code; added tachyon detector cap

Updated caps/deflectors to be multiplicative instead of additive; capped capacitor strength at 2.55x; capped tachyon detector cloak value at 81%

Bertie690 committed:
minor fixes

fixed an oopsie

Bertie690 committed:
Update traits.go

Added minimum decay stat to PRT stats

Bertie690 committed:
Revert "Update traits.go"

This reverts commit 35bdc9de71b8a13129fd6b3334e5abc15e6b59d6.

Bertie690 committed:
minor update to mineralpacket.go (more will come later when I finish stuff)

Bertie690 committed:
Update traits.go

fixed typo that undercounted PP permaform chance by ~20x

Bertie690 committed:
updated packet files

still gotta add the test func

Bertie690 committed:
Mineral Packet test changes

will open separate PR shortly - did oopsy and need to correct it

Bertie690 committed:
Added TestMineralPacket_estimateDamage test function

had to revert changes bc I accidentally discarded them

Bertie690 committed:
Update mineralpacket.go

Bertie690 committed:
IT packet decay fixing

turns out IT overflung packets decay even faster than normal (according to manual)

sirgwain commented on cs/traits.go:
```diff
@@ -282,7 +282,7 @@ func caSpec() PRTSpec {
 	spec.Instaforming = true
 	spec.PermaformChance = .1 // chance is 10% if pop is over 100k
-	spec.PermaformPopulation = 100_000
+	spec.PermaformPopulation = 100000
```
Can you revert this please? golang supports adding thousands separators to numeric literals.

sirgwain commented on cs/traits.go:
```diff
@@ -364,8 +364,8 @@ func ppSpec() PRTSpec {
-	spec.PacketTerraformChance = .5   // 50% per 100kT uncaught
-	spec.PacketPermaformChance = .001 // .1% per 100kT uncaught
+	spec.PacketTerraformChance = .5  // 50% per 100kT uncaught
+	spec.PacketPermaformChance = .05 // 5% per 100kT uncaught
```
This is supposed to be .1% per the manual.

Bertie690 replied:
huh - the autohost article you linked said it was 5%, so I just assumed that was the case without checking😅
(also the packet terraform functions should take into account the type of minerals received when doing terraforming checks)

sirgwain commented on cs/mineralpacket.go:
```diff
@@ -216,6 +214,53 @@ func (packet *MineralPacket) getDamage(rules *Rules, planet *Planet, planetPlaye
+	spd := packet.WarpSpeed * packet.WarpSpeed
+	totalDist := packet.Position.DistanceTo(target.Position)
+	decayRate := 0.0
+	ETA := int(math.Ceil(totalDist / float64(spd)))
```
should be lowercase `eta`

sirgwain commented on cs/mineralpacket.go:
```diff
@@ -216,6 +214,53 @@ func (packet *MineralPacket) getDamage(rules *Rules, planet *Planet, planetPlaye
+	decayRate := 0.0
+	ETA := int(math.Ceil(totalDist / float64(spd)))
+	//save copy of packet to revert to later
+	packetCopy := packet
```
Because `packet` is a pointer, this won't actually make a copy. Any changes will update the actual packet we are testing. You want:

```golang
packetCopy := *packet
```
And then have the function use `packetCopy` everywhere instead of `packet` so the original packet will be unchanged.

Bertie690 replied:
Would you need to delete packetCopy afterwards?
(and if so how would one go about that?)

sirgwain replied:
no, it will be cleaned up automatically when the function exits.

sirgwain commented on cs/mineralpacket_test.go:
```diff
@@ -134,3 +134,45 @@ func TestMineralPacket_getPacketDecayRate(t *testing.T) {
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			player := NewPlayer(1, tt.args.race).withSpec(&rules)
+			planet := NewPlanet().withPosition(tt.args.planetPosition)
```
```suggestion
			player := NewPlayer(1, tt.args.race).withSpec(&rules).WithNum(1)
			planet := NewPlanet().withPosition(tt.args.planetPosition).WithPlayerNum(1)
```

PlayerNums of `0` are considered unowned (because it's the default value in the struct). You have to make the planet know it's owned by a player and the player needs to have a number.

sirgwain commented on cs/mineralpacket_test.go:
```diff
@@ -134,3 +134,45 @@ func TestMineralPacket_getPacketDecayRate(t *testing.T) {
+			planet := NewPlanet().withPosition(tt.args.planetPosition)
+			planet.Spec.Population = tt.args.planetPop
+			planet.Spec.DefenseCoverage = tt.args.planetDefCoverage
+			planet.Spec.PlanetStarbaseSpec.SafePacketSpeed = tt.args.planetDriverSpeed
```
```suggestion
			planet.Spec.PlanetStarbaseSpec.SafePacketSpeed = tt.args.planetDriverSpeed
			planet.Spec.HasStarbase = true
```

sirgwain commented on cs/mineralpacket_test.go:
```diff
@@ -134,3 +134,45 @@ func TestMineralPacket_getPacketDecayRate(t *testing.T) {
+		args   args
+		want   MineralPacketDamage
+	}{
+		{"1 yr away; no decay/dmg", fields{WarpSpeed: 5, SafeWarpSpeed: 5}, args{NewRace().WithSpec(&rules), 5, Vector{25.0, 0.0}, 0.0, NewRace().WithSpec(&rules), 1000000, Cargo{10, 10, 10, 0}}, MineralPacketDamage{}},
```
This is a bit hard to know what each arg is, and you don't need `.0` on the floats since they are already declared as floats. Can you update the args to have named args?

```suggestion
		{
			"1 yr away; no decay/dmg",
			fields{WarpSpeed: 5, SafeWarpSpeed: 5},
			args{
				race:              NewRace().WithSpec(&rules),
				planetDriverSpeed: 5,
				planetPosition:    Vector{25, 0},
				planetDefCoverage: 0,
				targetRace:        NewRace().WithSpec(&rules),
				planetPop:         1000000,
				mass:              Cargo{10, 10, 10, 0},
			},
			MineralPacketDamage{},
		},
```

sirgwain commented on cs/mineralpacket_test.go:
```diff
@@ -134,3 +134,45 @@ func TestMineralPacket_getPacketDecayRate(t *testing.T) {
+		t.Run(tt.name, func(t *testing.T) {
+			player := NewPlayer(1, tt.args.race).withSpec(&rules)
+			planet := NewPlanet().withPosition(tt.args.planetPosition)
+			planet.Spec.Population = tt.args.planetPop
```
you'd think the `Spec.Population` would work, but I use that for the UI, not the code. I probably should remove it and just let the UI compute the pop based on the planet cargo.
```suggestion
			planet.setPopulation(tt.args.planetPop)
```

sirgwain reviewed (changes requested):
I made some code suggestions to fix the unit tests to run as you'd expect. Something is up with the decayRate on the estimator though. For the second test, it's not decaying the packet.

Bertie690 committed:
Fixed stuff

Hopefully it doesn't bork now lol (mark my words it will)

Bertie690 commented:
> I made some code suggestions to fix the unit tests to run as you'd expect. Something is up with the decayRate on the estimator though. For the second test, it's not decaying the packet.

After some debugging, I figured out why this wasn't working - I had set the `decayRate` variable to equal `1 - getPacketDecayRate()` when I should've just used `getPacketDecayRate()` instead. (This makes sense as to(This makes sense as to 
I also realized that the other main reason why my tests were breaking was because my math was off. It _should_ work now, but let me know if anything's still breaking.

Bertie690 committed:
Minor packet tweaks

Added another test func to confirm that min decay works; added minor shortcut to skip calculating decay if packet can be caught safely

sirgwain committed:
Merge branch 'develop' of github.com:sirgwain/craig-stars into packet

sirgwain commented on cs/turn.go:
```diff
@@ -1073,9 +1073,19 @@ func (t *turn) decaySalvage() {
-		packet.Cargo = packet.Cargo.Multiply(decayRate)
+
+		//loop through all 3 mineral types and reduce each one in turn
+		for _, minType := range [3]CargoType{Ironium, Boranium, Germanium} {
+			mineral := float64(packet.Cargo.GetAmount(minType))
+			decayAmount := MaxInt(int(decayRate * mineral), int(float64(t.game.Rules.PacketMinDecay)*float64(player.Race.Spec.PacketDecayFactor)))
+			packet.Cargo.SubtractAmount(minType, decayAmount)
+			if packet.Cargo.GetAmount(minType) < 1 {
+				packet.Cargo = packet.Cargo.MinZero()
+				packet.Cargo.AddAmount(minType, 1)
+			}
+		}
```
something is wrong with this function. I generated a test turn and all the packets in game were decayed to 1kT each.

Also, we want to allow decaying to 0 because that means the packet fades away to nothing.

Bertie690 replied:
oop
Probably need to remove the MinZero section...

sirgwain commented on cs/mineralpacket.go:
```diff
@@ -150,13 +149,20 @@ func (packet *MineralPacket) completeMove(rules *Rules, player *Player, planet *
 		packet.checkTerraform(rules, player, planet, damage.Uncaught)
 		packet.checkPermaform(rules, player, planet, damage.Uncaught)
+		receiverDriverSpeed := 0
+		if planet.Spec.HasStarbase {
+			receiverDriverSpeed = planet.Spec.SafePacketSpeed
+		}
+		percentCaughtSafely := float64((packet.WarpSpeed * packet.WarpSpeed) / (receiverDriverSpeed * receiverDriverSpeed))
```
```suggestion
		var percentCaughtSafely float64
		if planet.Spec.HasStarbase && planet.Spec.SafePacketSpeed > 0 {
			percentCaughtSafely = float64((packet.WarpSpeed * packet.WarpSpeed) / (planet.Spec.SafePacketSpeed * planet.Spec.SafePacketSpeed))
		}
```

this is causing a divide by zero error if the player has no receiver on their starbase.

Bertie690 committed:
Update cs/mineralpacket.go

Co-authored-by: sirgwain <sirgwain@users.noreply.github.com>

Bertie690 committed:
Update mineralpacket.go

Bertie690 committed:
Merge branch 'packet' of https://github.com/Bertie690/craig-stars into packet

Bertie690 committed:
Update turn.go

i forgot to change this earlier. 
Woe is me

Bertie690 committed:
Update mineralpacket.go

sirgwain commented on cs/turn.go:
```diff
@@ -1073,9 +1073,15 @@ func (t *turn) decaySalvage() {
-		decayRate := 1 - packet.getPacketDecayRate(&t.game.Rules, &player.Race)*(packet.distanceTravelled/float64(packet.WarpSpeed*packet.WarpSpeed))
-		packet.Cargo = packet.Cargo.Multiply(decayRate)
+		// update the decay amount based on this distance traveled this turn
+		decayRate := packet.getPacketDecayRate(&t.game.Rules, &player.Race)*(packet.distanceTravelled/float64(packet.WarpSpeed*packet.WarpSpeed))
```
One more fix, if the decayRate is 0, the packet shouldn't decay at all.
```suggestion
		decayRate := packet.getPacketDecayRate(&t.game.Rules, &player.Race) * (packet.distanceTravelled / float64(packet.WarpSpeed*packet.WarpSpeed))
		if decayRate == 0 {
			continue
		}
```

Bertie690 committed:
Update turn.go

Hopefully actually fixed it this time fr
(Also I put lines 1100-1105 (the debug log message for deleting packets) inside the conditional; if it turns out to screw stuff up feel free to undo it)

Bertie690 committed:
hopefully fixed test functions again

Bertie690 committed:
fixed tpyo

oopies daiys

Bertie690 committed:
Update intel.go

Updated mineral packet function to use new function estimateDamage instead of getDamage

Bertie690 committed:
Update mineralpacket.go

added special "exit code" status for if the packet will vanish before impact (easier than re-doing all the relevant calculations all over again)

Bertie690 committed:
Update message.go

Implemented new packet messages & new estimateDamage function

sirgwain committed:
Merge branch 'develop' into packet

sirgwain commented on cs/shipdesign.go:
```diff
@@ -348,9 +348,9 @@ func ComputeShipDesignSpec(rules *Rules, techLevels TechLevel, raceSpec RaceSpec
-			spec.BeamBonus += component.BeamBonus * float64(slot.Quantity)
-			spec.BeamDefense += component.BeamDefense * float64(slot.Quantity)
+			// beam bonuses (capacitors are capped at 2.55x; beam deflectors are unlimited)
+			spec.BeamBonus += math.Min(math.Pow(1+component.BeamBonus, float64(slot.Quantity))-1, 1.55)
```
shouldn't this be 2.55 not 1.55?

Bertie690 replied:
Not to my knowledge - IIRC the beam bonus gets added on to the beam's power (total power multi is 1+beamBonus), so 2.55 would actually result in a max boost of 3.55x.

Bertie690 committed:
Update message.go

}

Bertie690 committed:
Bugfixes

yay

Bertie690 committed:
minor fixes

Estimated damage test #2 ("3 lvls overwarp + 1 yr travel w/ min decay (300 dmg)") is still breaking and I have no idea why
